fuel-vm/src/interpreter/flow.rs

```diff
@@ -149,138 +161,385 @@ where
-        }
+        *self.registers.read_registers.sp += len;
+        *self.registers.read_registers.ssp = *self.registers.read_registers.sp;
 
-        self.registers[RegId::BAL] = b;
-        self.registers[RegId::PC] = fpx;
-        self.registers[RegId::IS] = self.registers[RegId::PC];
-        self.registers[RegId::CGAS] = forward_gas_amount;
+        let code_mem_range = CheckedMemRange::new(*self.registers.read_registers.fp, len as usize)?;
+        let frame_end = write_call_to_memory(&frame, frame_bytes, code_mem_range, self.memory.memory, self.storage)?;
+        *self.registers.read_registers.bal = self.params.amount_of_coins_to_forward;
+        *self.registers.read_registers.pc = frame_end;
+        *self.registers.read_registers.is = *self.registers.read_registers.pc;
+        *self.registers.read_registers.cgas = forward_gas_amount;
 
         let receipt = Receipt::call(
             id,
             *frame.to(),
-            b,
+            self.params.amount_of_coins_to_forward,
             *frame.asset_id(),
-            d,
+            self.params.amount_of_gas_to_forward,
             frame.a(),
             frame.b(),
-            self.registers[RegId::PC],
-            self.registers[RegId::IS],
+            *self.registers.read_registers.pc,
+            *self.registers.read_registers.is,
         );
 
-        self.append_receipt(receipt);
+        append_receipt(self.receipts, self.script, self.consensus, self.memory.memory, receipt);
 
         self.frames.push(frame);
 
         Ok(())
     }
+}
 
-    /// Prepare a call instruction for execution
-    pub fn prepare_call(&mut self, ra: RegId, rb: RegId, rc: RegId, rd: RegId) -> Result<(), RuntimeError> {
-        const M: &str = "the provided id is not a valid register";
+fn write_call_to_memory<S>(
+    frame: &CallFrame,
+    frame_bytes: Vec<u8>,
+    code_mem_range: CheckedMemRange,
+    memory: &mut [u8; VM_MEMORY_SIZE],
+    storage: &S,
+) -> Result<Word, RuntimeError>
+where
+    S: StorageSize<ContractsRawCode> + StorageRead<ContractsRawCode> + StorageAsRef,
+    <S as StorageInspect<ContractsRawCode>>::Error: Into<std::io::Error>,
+{
+    let mut code_frame_range = code_mem_range.clone();
+    // Addition is safe because code size + padding is always less than len
+    code_frame_range.shrink_end((frame.code_size() + frame.code_size_padding()) as usize);
+    code_frame_range.clone().write(memory).copy_from_slice(&frame_bytes);
+
+    let mut code_range = code_mem_range.clone();
+    code_range.grow_start(CallFrame::serialized_size());
+    code_range.shrink_end(frame.code_size_padding() as usize);
+    let bytes_read = storage
+        .storage::<ContractsRawCode>()
+        .read(frame.to(), code_range.write(memory))
+        .map_err(RuntimeError::from_io)?
+        .ok_or(PanicReason::ContractNotFound)?;
+    if bytes_read as Word != frame.code_size() {
+        return Err(PanicReason::ContractNotFound.into());
```

Comment:
It would be nice to have another error here

Reply:
What kind of error are you thinking?

Reply:
Like `ContractSizeMismatch`